### DISCUSSION TITLE
share clipboard can't work properly

### DISCUSSION THREAD

**Discussion by gbcbooks** [rustdesk/rustdesk#9105] [General]
### Bug Description

I've heard of v1.2.7 support copy image already.  
https://github.com/rustdesk/rustdesk/discussions/8467
https://x.com/rustdesk/status/1807981430775206289

from my test, copy text or image (screenshot), sometime it works, but something it doesn't. 

### How to Reproduce

sometime it works, but something it doesn't. 

### Expected Behavior

fix the issue and improve user experience

### Operating system(s) on local side and remote side

Ubuntu 24.04 -> Windows 11

### RustDesk Version(s) on local side and remote side

1.2.7 -> 1.2.7

### Screenshots

![rustdesk-clipboard](https://github.com/user-attachments/assets/4820f6dd-21ea-4511-8cb7-dffe6885d7ed)

**Comment by fufesou** [comment]
There's a bug of sharing images for windows installed version 1.2.7.

Please try the nightly build.

**Reply by gbcbooks** [reply]
actully , I use nightly build before today.  I just upgrade from nightly v1.2.7 to v1.2.7. they have the same issue.

**Reply by fufesou** [reply]
Hi, can you help confirm this again?
1. Text copy and paste sometimes also doesn't work?
2. Does the controlled computer have multiple remote control software?
3. Is the image copy and paste slow? Can it be pasted after 1 second?

**Reply by gbcbooks** [reply]
1. Yes, it is
2. Yes , I install anydesk on it.  It is just running in the background, and remote control haven't been activated
3. No, It wasn't slow when this function is good. and It can't be pasted when this function is not good.

**Reply by fufesou** [reply]
Thansk for your kind feedback.

I've tested Ubuntu 22.04 -> Win10, both Wayland(Gnome) and X11 are tested.
It's a little hard for me to reproduce this issue.

1. Do you use Wayland(Gnome) or X11? 
2. How is this problem more likely to be triggered? Did you copy something that is likely to trigger it? Or how long after you use it does this problem get triggered?

Thanks

**Reply by gbcbooks** [reply]
thank you for giving time to my issue

1. I use X11.
2. I am not sure, But I can list the softwares i use which is suspicious can cause this issue. 
    2.1. Remmina, i use  its spice viewer remote the same win11(It is my VM)
    2.2. flameshot,  for the screenshot
    2.3. Anydesk, for the backup remote softwares, But it didnot start when.

**Reply by gbcbooks** [reply]
and 2.4. KeepassXC , this software also use clipboard too. my setting is 150s clear the clipboard if copy password from it.

**Reply by gbcbooks** [reply]
Can you tell me which udp port will be used to share the clipboard content ?

**Reply by fufesou** [reply]
The clipboard content is not shared with UDP.

It uses message transport and uses the same TCP module as other messages.

**Reply by fufesou** [reply]
> sometime it works, but something it doesn't.

Can you confirm if `1.2.7` also sometimes fails to copy text? Or can you please spend more time on `1.2.7` to confirm it?

`1.2.7` did not change the strategy of copying text.

If it also has this problem, it can basically be confirmed that there is a conflict between multiple software using the pasteboard.

Anyway, I will continue to do more testing.

**Reply by gbcbooks** [reply]
Currently,  I tested again,  copying text is fine, But image is having this issue. 
![rustdesk-clipboard-01](https://github.com/user-attachments/assets/0b9b04aa-cc0f-42b2-b275-915180f088cb)

**Reply by fufesou** [reply]
> Currently, I tested again, copying text is fine, But image is having this issue.

Thanks.

`1.2.7` installed version, on Windows, does have this issue. 

So

`1.2.7` text copy&paste works fine. `1.3.0` copy&paste (both text and image) has this issue?

**Reply by gbcbooks** [reply]
> > sometime it works, but something it doesn't.
> 
> Can you confirm if `1.2.7` also sometimes fails to copy text? Or can you please spend more time on `1.2.7` to confirm it?
> 
> `1.2.7` did not change the strategy of copying text.
> 
> If it also has this problem, it can basically be confirmed that there is a conflict between multiple software using the pasteboard.
> 
> Anyway, I will continue to do more testing.

If so , Will it used tcp port 21114 ? Cause i found many tcp retransmision to tcp ort 21114.  I did see the statement mention it use e this tcp port.

**Reply by fufesou** [reply]
> Will it used tcp port 21114

No. `21114` is the port of the http server. For self-hosting users.

**Reply by gbcbooks** [reply]
I have time , do you want to use my environment to test it?

**Reply by fufesou** [reply]
Thanks. This is my email `bGlubG9uZzEyNjZAZ21haWwuY29t` (in base64). Please send me an email.

**Reply by gbcbooks** [reply]
I might find out the trigger,  if I use ubuntu's orignal screenshot tool,  i can copy image now.  if i use flameshot, it doesn't work, But the strange thing is that the screenshots I took with flameshot can be used in other software.

**Reply by gbcbooks** [reply]
> Thanks. This is my email `bGlubG9uZzEyNjZAZ21haWwuY29t` (in base64). Please send me an email.

sent

**Reply by fufesou** [reply]
> if I use ubuntu's orignal screenshot tool, i can copy image now. if i use flameshot, it doesn't work

Please double confirm.

The text copy&paste always work?

> But the strange thing is that the screenshots I took with flameshot can be used in other software.

It may be related to the image type in the pasteboard. RustDesk only read `svg` and `png` formats.

https://github.com/rustdesk-org/arboard/blob/a04bdb1b368a99691822c33bf0f7ed497d6a7a35/src/platform/linux/x11.rs#L939

**Reply by gbcbooks** [reply]
> > if I use ubuntu's orignal screenshot tool, i can copy image now. if i use flameshot, it doesn't work
> 
> Please double confirm.
> 
> The text copy&paste always work?
> 
> > But the strange thing is that the screenshots I took with flameshot can be used in other software.
> 
> It may be related to the image type in the pasteboard. RustDesk only read `svg` and `png` formats.
> 
> https://github.com/rustdesk-org/arboard/blob/a04bdb1b368a99691822c33bf0f7ed497d6a7a35/src/platform/linux/x11.rs#L939

currently, yes ,  text copy& paste always work. 

and , my flameshot setting is saving type of png now.  you can confirm from here. 
![image](https://github.com/user-attachments/assets/ffe6697e-85ce-4918-a935-d5a23bc0bdd2)

**Comment by fufesou** [comment]
The RustDesk clipboard listener needs improvement.

Linux (x11) -> Windows

1. Connect RustDesk. (Windows 1)
2. Connect Remmina. (Windows 2)
3. Disconnect Remmina.
4. Then clipboard, Linux -> Windows is broken, the clipboard listener does not work any more.

Reconnecting will not fix the clipboard listener.
The service must be restarted.

**Comment by rafitamolin** [comment]
I have a strange bug which has been occurring since few versions ago, and stills occurs on latest nightly, which may be related with this issue, @fufesou.

Two Windows desktops connected with Rustdesk latest nightly. The "Allow copy pasting files" option is disabled and the "Disable clipboard" option is enabled. It also occurs with the "Disable clipboard" option disabled.

That causes that the local computer's clipboard doesn't 90% of the time. If you copy text, it doesn't copy anything. With images, something it copies an empty file and other times nothing is copied. The remote computer's clipboard is fine.

When closing the remote connection, the local computer's cliboard is working perfectly fine.

**Reply by fufesou** [reply]
Thanks for your feedback.

Do you use the latest nightly on the three PCs?
Do you have multiple Romote apps running?

> Two Windows desktops connected with Rustdesk latest nightly. 

There's no difference whether it's one or two controlled side. There is only one clipboard listener.

I cannot reproduce your issue. There may be something different.

**Reply by rafitamolin** [reply]
It's actually only two PCs, one local and one remote. I only have RustDesk installed as my remote app in both PCs.

I'm using a ShareX (screen capturing software) hotkey (Ctrl + Impr Pant) which opens a region capturer (rectangle) and then after selecting that rectangle in the screen, the screenshot is automatically copied to the clipboard.

That works fine when in my local PC there's no remote connection opened on RustDesk, but when there is, it fails 90% of the time to copy it.

I have now tried copying normal text and capturing screenshots with the Snipping Tool pre-installed Windows app and it works fine. Maybe ShareX is interfering in some way with RustDesk in the clipboard listener.

**Reply by fufesou** [reply]
Sorry, I still cannot reproduce your issue.

I've tested `ShareX 16.1` (Installed and portabla) and `RustDesk 1.3.0` (Installed), Win11 as the control side.

But I agree that "ShareX" might conflict with RustDesk on the clipboard.

I will continue testing and fixing it after I finish the higher priority tasks at hand.

If you have more clues, please post them here, thank you.
